src/charm_state.py

```diff
@@ -0,0 +1,171 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module defining the CharmState class which represents the state of the SMTP Integrator charm."""
+
+import itertools
+from typing import Optional
+
+import ops
+from charms.smtp_integrator.v0 import smtp
+from pydantic import BaseModel, Field, ValidationError
+
+KNOWN_CHARM_CONFIG = (
+    "host",
+    "port",
+    "user",
+    "password",
+    "auth_type",
+    "transport_security",
+    "domain",
+)
+
+
+class SmtpIntegratorConfig(BaseModel):  # pylint: disable=too-few-public-methods
+    """Represent charm builtin configuration values.
+
+    Attrs:
+        host: The hostname or IP address of the outgoing SMTP relay.
+        port: The port of the outgoing SMTP relay.
+        user: The SMTP AUTH user to use for the outgoing SMTP relay.
+        password: The SMTP AUTH password to use for the outgoing SMTP relay.
+        auth_type: The type used to authenticate with the SMTP relay.
+        transport_security: The security protocol to use for the outgoing SMTP relay.
+        domain: The domain used by the sent emails from SMTP relay.
+    """
+
+    host: str = Field(..., min_length=1)
+    port: int
+    user: Optional[str]
+    password: Optional[str]
+    auth_type: Optional[smtp.AuthType]
+    transport_security: Optional[smtp.TransportSecurity]
+    domain: Optional[str]
+
+
+class CharmConfigInvalidError(Exception):
+    """Exception raised when a charm configuration is found to be invalid.
+
+    Attrs:
+        msg (str): Explanation of the error.
+    """
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the CharmConfigInvalidError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+class CharmState:
```

Comment:
Recently I had a review on Synapse to make CharmState a dataclass. What do you think?

Reply:
Makes sense